File: problems/0005_longest-palindromic-substring/central-expansion/Solution.swift
class Solution {
    func longestPalindrome(_ s: String) -> String {
        let (chars, sLen) = (s.utf8CString, s.count)
        var (maxLeft, maxRight) = (0, 0)
        for i in 0...(sLen * 2) {
            var posLeft = i / 2
            var posRight = (i % 2 == 1 ? posLeft + 1 : posLeft)
            while 0 <= posLeft && posRight < sLen &&
                    chars[posLeft] == chars[posRight] {
                posLeft -= 1
                posRight += 1
            }
            posLeft += 1
            if posRight - posLeft > maxRight - maxLeft {
                (maxLeft, maxRight) = (posLeft, posRight)
            }
        }
        return substring(s, maxLeft, maxRight)
    }
    @inline
    func substring(_ s: String, _ start: Int, _ end: Int) -> String {
        return String(s[s.index(s.startIndex, offsetBy: start)..<s.index(s.startIndex, offsetBy: end)])
    }
}

File: problems/0005_longest-palindromic-substring/central-expansion/Solution.cs
public class Solution {
    public string LongestPalindrome(string s) {
        int iEnd = s.Length * 2 - 1,
            start = 0, maxLen = 0;
        for (int i = 0; i < iEnd; ++i) {
            int posLeft = i / 2,
                posRight = (i % 2 == 1 ? posLeft + 1 : posLeft);
            while (0 <= posLeft && posRight < s.Length &&
                    s[posLeft] == s[posRight]) {
                --posLeft;
                ++posRight;
            }
            int len = posRight - (++posLeft);
            if (len > maxLen) {
                start = posLeft;
                maxLen = len;
            }
        }
        return s.Substring(start, maxLen);
    }
}
